pkg/client/core/service_monitoring_client.go

```diff
@@ -40,28 +50,206 @@ const (
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = s.spinner.Stop()
+	}()
+
+	chart := promtail.New(nil)
+
+	err = s.api.DeletePromtail(api.DeleteHelmReleaseOpts{
+		ID:          opts.ID,
+		ReleaseName: chart.ReleaseName,
+		Namespace:   chart.Namespace,
+	})
+	if err != nil {
+		return err
+	}
+
+	report, err := s.store.RemovePromtail(opts.ID)
+	if err != nil {
+		return err
+	}
+
+	return s.report.ReportRemovePromtail(report)
+}
+
+func (s *monitoringService) CreatePromtail(_ context.Context, opts client.CreatePromtailOpts) (*client.Promtail, error) {
+	err := s.spinner.Start("promtail")
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err = s.spinner.Stop()
+	}()
+
+	chart, err := s.api.CreatePromtail(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	l := &client.Promtail{
+		ID:    opts.ID,
+		Chart: chart,
+	}
+
+	report, err := s.store.SavePromtail(l)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.report.ReportSavePromtail(l, report)
+	if err != nil {
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (s *monitoringService) DeleteLoki(ctx context.Context, opts client.DeleteLokiOpts) error {
+	err := s.spinner.Start("loki")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = s.spinner.Stop()
+	}()
+
+	chart := loki.New(nil)
+
+	err = s.manifest.DeleteNativeSecret(ctx, client.DeleteNativeSecretOpts{
+		ID:        opts.ID,
+		Name:      lokiDatasourceSecretName,
+		Namespace: config.DefaultMonitoringNamespace,
+	})
+
+	err = s.api.DeleteLoki(api.DeleteHelmReleaseOpts{
+		ID:          opts.ID,
+		ReleaseName: chart.ReleaseName,
+		Namespace:   chart.Namespace,
+	})
+	if err != nil {
+		return err
+	}
+
+	report, err := s.store.RemoveLoki(opts.ID)
 	if err != nil {
 		return err
 	}
 
+	return s.report.ReportRemoveLoki(report)
+}
+
+func (s *monitoringService) CreateLoki(ctx context.Context, opts client.CreateLokiOpts) (*client.Loki, error) {
+	err := s.spinner.Start("loki")
+	if err != nil {
+		return nil, err
+	}
+
 	defer func() {
 		err = s.spinner.Stop()
 	}()
 
-	// This is too heavy handed, we should only remove the chart
-	// and the secrets manifest. We can, however, wait until Loki
-	// with doing this.
-	err = s.manifest.DeleteNamespace(ctx, api.DeleteNamespaceOpts{
+	chart, err := s.api.CreateLoki(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := yaml.Marshal(lokipkg.NewDatasourceTemplate())
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = s.manifest.CreateNativeSecret(ctx, client.CreateNativeSecretOpts{
 		ID:        opts.ID,
+		Name:      lokiDatasourceSecretName,
 		Namespace: config.DefaultMonitoringNamespace,
+		Data: map[string]string{
+			"loki-datasource.yaml": string(data),
+		},
+		Labels: map[string]string{
+			"grafana_datasource": "1",
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// The datasources are only loaded during grafana startup, so we need
+	// to cycle grafana to have it pick up the changes
+	for _, replicas := range []int32{0, 1} {
```

Comment:
Can we be sure that the ordering is preserved? I.e. that what k8s actually performs is

```shell
kubectl scale --replicas=0
kubectl scale --replicas=1
```

and not vice versa?

I'm guessing that this most likely is not a problem, but I was unable to google an answer to this.

Reply:
Great question, you can read about it here: https://golang.org/ref/spec#For_range
> For an array, pointer to array, or slice value a, the index iteration values are produced in increasing order, starting at element index 0.

So in this case, the order is given and will always be the same. For maps, however, the order is not given, and may be different for each iteration.

Reply:
Ah, this was more an k8s API question, not a Golang question. Do we know that the k8s api server will perform to operations from two rapidly incoming api calls in the order the api calls came?

Reply:
Aha, so they are blocking calls, so I hope so, but now you made me nervous 🤔 I will make a separate PR where we watch for the expected result before returning

Reply:
Very valid question